### PR TITLE
Multi user support

### PR DESCRIPTION
**Pull request by GauthamGoli**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?
- [x] Have you successfully run `brew man` locally and committed any changes?

-----

To be merged along with https://github.com/Homebrew/install/pull/328

### CODE REVIEW

File: Library/Homebrew/keg.rb
Comment (reitermarkus) [inline Library/Homebrew/keg.rb:112]: No need for the parentheses here.

File: Library/Homebrew/keg.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/keg.rb:107]: Would be good to use `env_config.rb` for this.

File: Library/Homebrew/keg.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/keg.rb:108]: What happens if an install is attempted and these directories don't exist or aren't writable?

File: Library/Homebrew/lock_file.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/lock_file.rb:46]: Would be good to use `env_config.rb` for this.

**Review by MikeMcQuaid — COMMENTED** [review]
Nice work here @GauthamGoli! 👏🏻

File: Library/Homebrew/keg.rb
Comment (GauthamGoli) [reply Library/Homebrew/keg.rb:108]: @MikeMcQuaid 
- If directories do not exist:  They are created and set with `755` octal permissions  (https://github.com/Homebrew/install/blob/master/install.sh#L402)

- If directories exit but are not writable:  Permissions are updated to `755` 
(https://github.com/Homebrew/install/blob/master/install.sh#L414)

Comment (MikeMcQuaid) [reply Library/Homebrew/keg.rb]: Sounds good!

File: Library/Homebrew/keg.rb
Comment (reitermarkus) [inline Library/Homebrew/keg.rb:111]:
```suggestion
  ZSH_DIRECTORIES = if Homebrew::EnvConfig.multi_user?
    []
  else
    %w[share/zsh share/zsh/site-functions].map { |dir| HOMEBREW_PREFIX/dir }
  end.freeze
```

**Review by MikeMcQuaid — APPROVED** [review]

File: Library/Homebrew/env_config.rb
Comment (EricFromCanada) [inline Library/Homebrew/env_config.rb:202]:
```suggestion
        description: "If set, enable multi-user support. Allows users belonging to `admin` group to " \
                     "install formulae.",
```

**Comment by robsonpeixoto** [comment]
Thanks man! I'll need it this next week. ❤️
I'll need to keep the envvar `HOMEBREW_MULTI_USER` on my .basrc/zshrc/... after install homebrew or is only needed on the install.

**Comment by GauthamGoli** [comment]
@robsonpeixoto The feature is still a work in progress, setting the ENV var will not make it work at the moment, there's another PR that needs to be taken care of https://github.com/Homebrew/install/pull/328

**Comment by gromgit** [comment]
Suggest adding near the top of `bin/brew`:
```
# Force a sane umask for multi-user support
if [[ -n "${HOMEBREW_MULTI_USER-}" ]]; then
  umask 002
fi
```
otherwise various "permission denied" errors may creep in over time.

**Comment by alexreg** [comment]
Glad to see efforts to integrate multi-user support into Homebrew itself. Is this intended to operate in a similar way to [brewdo](https://github.com/brewdo/brewdo)? More or less, that means having a `_homebrew` user (and group) that takes ownership of `/usr/local/{Homebrew,Cellar,Caskroom}` as well as all the Homebrew-installed files in `/usr/local.` (but not existing directories/files)? And naturally, doing something like `sudo -u _homebrew -g _homebrew ...` for all commands that make modifications to `/usr/local/`.

**Comment by gromgit** [comment]
> Is this intended to operate in a similar way to [brewdo](https://github.com/brewdo/brewdo)?

No. This operates along the lines of "users in `admin` group can manage Homebrew directly, without `sudo`".

**Comment by alexreg** [comment]
I presume there is still a special `_homebrew` *user* or such though?

**Comment by gromgit** [comment]
> I presume there is still a special `_homebrew` _user_ or such though?

No. That's why I [made my `umask` recommendation], to ensure that `brew` doesn't mask out group write bits even if the running user does so.

**Comment by alexreg** [comment]
Hmm, I'm not sure that's ideal still, but I see your rationale. (I may decide to keep maintaining brewdo in this case.) Thanks for clarifying, regardless.

**Comment by MikeMcQuaid** [comment]
@alexreg My suspicion is that this will make things easier for configurations like `brewdo` while not removing the need for `brewdo` entirely. Feel free to open your own PRs to move things in this direction as desired; don't be blocked on this one if you've got an itch to improve things.

**Comment by alexreg** [comment]
@MikeMcQuaid Good to know, thanks. And you're right, looking at these changes properly now, it won't replicate brewdo's behaviour, but it may make it work slightly smoother, and certainly shouldn't create any incompatibilities.

Is there definitely a no-go on even small changes to the `install.sh` script, however? (Per that issue I created the other day.)

**Comment by MikeMcQuaid** [comment]
> Is there definitely a no-go on even small changes to the `install.sh` script, however? (Per that issue I created the other day.)

@alexreg Will review any PR on a case-by-case basis. We're not wanting to support more "options" in the controller or conditionally disable behaviour, though.

File: Library/Homebrew/env_config.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/env_config.rb:219]:
```suggestion
        description: "If set, formulae are installed as group-writable (default: user-writable).",
```

**Review by MikeMcQuaid — COMMENTED** [review]
One suggested wording change but otherwise happy with this. I'd suggest that everywhere we have `writable?` or `writable_real?` checks may need to be updated to use `group_writable?` or `group_writable_real?` (certainly in `diagnostic.rb` and `pathname.rb`).

**Review by twocountriestrus — APPROVED** [review]

**Comment by GauthamGoli** [comment]
Hi, all. I am closing this PR due to my inactivity on this. 

Anyone else is welcome to continue the work or start afresh. HMU in case of questions.   

Thanks to the maintainers for spending time on reviewing the PR! (Sorry, that it did not work out as expected)

**Pull request closed without merging**
